core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeService.java

```diff
@@ -32,26 +35,52 @@
     @Inject
     private StackUpdater stackUpdater;
 
-    public void upgradeClusterManager(long stackId, StatedImage targetImage) {
+    public boolean upgradeClusterManager(long stackId, StatedImage currentImage, StatedImage targetImage) {
+        String currentCmBuildNumber = currentImage.getImage().getCmBuildNumber();
+        boolean clusterManagerUpdateNeeded = !currentCmBuildNumber.equals(targetImage.getImage().getCmBuildNumber());
         flowMessageService.fireEventAndLog(stackId, Status.UPDATE_IN_PROGRESS.name(), DATALAKE_UPGRADE, targetImage.getImage().getUuid());
         clusterService.updateClusterStatusByStackId(stackId, Status.UPDATE_IN_PROGRESS);
-        flowMessageService.fireEventAndLog(stackId, Status.UPDATE_IN_PROGRESS.name(), CLUSTER_MANAGER_UPGRADE);
+        if (clusterManagerUpdateNeeded) {
+            flowMessageService.fireEventAndLog(stackId, Status.UPDATE_IN_PROGRESS.name(), CLUSTER_MANAGER_UPGRADE);
+        } else {
+            flowMessageService.fireEventAndLog(stackId, Status.UPDATE_IN_PROGRESS.name(), CLUSTER_MANAGER_UPGRADE_NOT_NEEDED, currentCmBuildNumber);
+        }
+        return clusterManagerUpdateNeeded;
     }
 
-    public void upgradeCluster(long stackId, Image image) {
-        String clusterManagerVersion = image.getVersion();
+    public boolean upgradeCluster(long stackId, StatedImage currentImage, StatedImage targetImage) {
+        Image targetIm = targetImage.getImage();
+        Image currentIm = currentImage.getImage();
+        String currentRuntimeBuildNumber = currentIm.getStackDetails().getStackBuildNumber();
+        boolean clusterManagerUpdateNeeded = !currentIm.getCmBuildNumber().equals(targetIm.getCmBuildNumber());
+        boolean clusterRuntimeUpgradeNeeded = !currentRuntimeBuildNumber.equals(targetIm.getStackDetails().getStackBuildNumber());
 
-        flowMessageService.fireEventAndLog(stackId, Status.UPDATE_IN_PROGRESS.name(), CLUSTER_MANAGER_UPGRADE_FINISHED, clusterManagerVersion);
+        if (clusterManagerUpdateNeeded) {
```

Comment:
could you pls add some extra logging here for the different clauses?

Reply:
The flowMessageService.fireEventAndLog() method will log the different kind of messages. I think it should be enough. What do you think?